TELCODOCS-1215 - Adding GitOps ZTP managed cluster maximums info (#60615)
openshift/openshift-docs: modules/ztp-gitops-ztp-max-spoke-clusters.adoc

Different versions of {rh-rhacm} can have varying performance characteristics and resource requirements.
Additionally, the configuration settings of {rh-rhacm}, such as the number of concurrent reconciliations or the frequency of health checks, can affect the managed cluster capacity of the hub cluster.

Use the following representative configuration and network specifications to develop your own Hub cluster and network specifications.

[IMPORTANT]
====
The following guidelines are based on internal lab benchmark testing only and do not represent a complete real-world host specification.
====

.Representative three-node hub cluster machine specifications
[cols=2*, width="90%", options="header"]
|====
|Requirement
|Description

|{product-title} version
|version 4.13

|{rh-rhacm} version
|version 2.7

|{cgu-operator-first}
|version 4.13

|Server hardware
|3 x Dell PowerEdge R650 rack servers

|NVMe hard disks
a|* 50 GB disk for `/var/lib/etcd`
* 2.9 TB disk for `/var/lib/containers`

|SSD hard disks
a|* 1 SSD split into 15 200GB thin-provisioned logical volumes provisioned as `PV` CRs
* 1 SSD serving as an extra large `PV` resource

|Number of applied DU profile policies

> |5 (6 if you include a validation policy)

Author: @akrzos Look OK to you?

Reviewer: No AIUI, it should be just 5 policies total. This is because there are 3 policies plus a site policy and a du validation policy.  In the scale lab tests I rename two parts of the PGT to match this same 5 policy count.

// Module included in the following assemblies:
//
// * scalability_and_performance/ztp_far_edge/ztp-preparing-the-hub-cluster.adoc]

:_content-type: REFERENCE
[id="ztp-gitops-ztp-max-spoke-clusters_{context}"]
= {ztp} managed cluster limits

With {ztp-first}, you can manage thousands of {sno} clusters in geographically displaced regions and networks.
The scaling limits for the number of clusters that can be managed from a single hub cluster will vary depending on various factors such as the following:

* Workload requirements, for example memory and network resources needed to support the expected workload demands

* Compute layer requirements for running the workloads, for example the number of `ReplicaSet`, `Deployment`, and `DaemonSet` resources that you require

* Container networking requirements, for example NFV, network policies, inter-pod communication settings, custom SDN solutions, and so on

> * Storage IOPS factors, such as whether the managed cluster host uses NVMe

Author: @imiller0 Oops! Friday brain :sweat_smile: Updated - PTAL.

Larger managed clusters with more nodes, namespaces, and resources require additional processing and management overhead.
Similarly, clusters with complex configurations or diverse workloads can require more resources from the hub cluster.

Number of managed policies::
The number of policies managed by the hub cluster scaled over the number of managed clusters bound to those policies is an important factor that determines how many clusters can be managed.

Monitoring and management workloads::
{rh-rhacm} continuously monitors and manages the managed clusters.
The number and complexity of monitoring and management workloads running on the hub cluster can affect its capacity.
Intensive monitoring or frequent reconciliation operations can require additional resources, potentially limiting the number of manageable clusters.

{rh-rhacm} version and configuration::
Different versions of {rh-rhacm} can have varying performance characteristics and resource requirements.
Additionally, the configuration settings of {rh-rhacm}, such as the number of concurrent reconciliations or the frequency of health checks, can impact the managed cluster capacity of the hub cluster.

> For details of managed {sno} cluster scale results performed by the Red Hat Performance and Scale lab, see link:https://access.redhat.com/documentation/en-us/red_hat_advanced_cluster_management_for_kubernetes/2.7/html/install/installing#single-node[Creating and managing {sno} clusters with {rh-rhacm}].

Author: I agree that the linked doc isn't very clear. 

I've reworded this section slightly, and made the managed clusters maximums claim explicit, with a footnote link to the source RHACM docs. @imiller0 WDYT? 

I think this latest version meets the original goal of documenting hub cluster maximums while also adding some extra context to set expectations for customers in real world situations.

> For details of managed {sno} cluster scale results performed by the Red Hat Performance and Scale lab, see link:https://access.redhat.com/documentation/en-us/red_hat_advanced_cluster_management_for_kubernetes/2.7/html/install/installing#single-node[Creating and managing {sno} clusters with {rh-rhacm}].

Reviewer 2: In my opinion it's pointless to give a date for a specific scale-test but not the _details_ on what was tested. 

Especially if we could provide the details of that scale test, relating to each one of those various considerations you listed and explained, I would feel this would be very valuable.

(If it's not clear, I really like the list of considerations you've produced :+1: )

// Module included in the following assemblies:
//
// * scalability_and_performance/ztp_far_edge/ztp-preparing-the-hub-cluster.adoc]

:_content-type: REFERENCE
[id="ztp-gitops-ztp-max-spoke-clusters_{context}"]
= {ztp} managed cluster limits

With {ztp-first}, you can manage thousands of clusters in geographically displaced regions and networks.
The Red Hat Performance and Scale lab successfully created and managed 3500 {sno} clusters from a single {rh-rhacm-first} hub cluster in a lab environment on 4 January 2023.footnote:[link:https://access.redhat.com/documentation/en-us/red_hat_advanced_cluster_management_for_kubernetes/2.7/html/install/installing#single-node[Creating and managing {sno} clusters with {rh-rhacm}]]

[IMPORTANT]
====
In real world situations, the scaling limits for the number of clusters that you can manage with a single hub cluster will vary depending on various factors affecting the hub cluster.

> ====

Reviewer 2: This warning looks a bit out-of-place.
Below you make it sufficiently clear that there are several factors to consider.
I don't feel there is any additional need for a big red warning sign

// Module included in the following assemblies:
//
// * scalability_and_performance/ztp_far_edge/ztp-preparing-the-hub-cluster.adoc]

:_content-type: REFERENCE
[id="ztp-gitops-ztp-max-spoke-clusters_{context}"]
= Recommended hub cluster specifications and managed cluster limits for {ztp}

> With {ztp-first}, you can manage thousands of clusters in geographically displaced regions and networks.

Reviewer 3: displaced or dispersed?

> The Red Hat Performance and Scale lab successfully created and managed 3500 virtual {sno} clusters from a single {rh-rhacm-first} hub cluster in a lab environment on 4 January 2023.footnote:[link:https://access.redhat.com/documentation/en-us/red_hat_advanced_cluster_management_for_kubernetes/2.7/html/install/installing#single-node[Creating and managing {sno} clusters with {rh-rhacm}]]

Reviewer 2: Is the date really relevant here for an external reader?

In real world situations, the scaling limits for the number of clusters that you can manage will vary depending on various factors affecting the hub cluster.
For example:

Hub cluster resources::
Available hub cluster host resources (CPU, memory, storage) are an important factor in determining how many clusters the hub cluster can manage.
The more resources allocated to the hub cluster, the more managed clusters it can accommodate.

Hub cluster storage::
The hub cluster host storage IOPS rating and whether the hub cluster hosts use NVMe storage can affect hub cluster performance and the number of clusters it can manage.

Network bandwidth and latency::
Slow or high-latency network connections between the hub cluster and managed clusters can impact how the hub cluster manages multiple clusters.

Managed cluster size and complexity::
The size and complexity of the managed clusters also affects the capacity of the hub cluster.
Larger managed clusters with more nodes, namespaces, and resources require additional processing and management overhead.

> Similarly, clusters with complex configurations or diverse workloads can require more resources from the hub cluster.

Reviewer 2: I feel this concern was still not addressed in the table below.

In my opinion, we should point out that it was the DU profile which is described in this section which the 3500 Nodes number refer to

{rh-rhacm} continuously monitors and manages the managed clusters.
The number and complexity of monitoring and management workloads running on the hub cluster can affect its capacity.
Intensive monitoring or frequent reconciliation operations can require additional resources, potentially limiting the number of manageable clusters.

{rh-rhacm} version and configuration::
Different versions of {rh-rhacm} can have varying performance characteristics and resource requirements.
Additionally, the configuration settings of {rh-rhacm}, such as the number of concurrent reconciliations or the frequency of health checks, can impact the managed cluster capacity of the hub cluster.

Use the following representative configuration and network specfications to develop your own Hub cluster and network specifications.

[IMPORTANT]
====
The following guidelines are based on internal benchmark testing only and do not represent a complete real-world host specification.
====

.Representative three-node hub cluster machine specifications
[cols=2*, width="90%", options="header"]
|====
|Requirement
|Description

|{product-title} version
|version 4.13

|{rh-rhacm} version

> |version {rh-rhacm-version}

Reviewer 2: There is a mismatch here, in the preview this renders to 2.8 while the section above states 2.7 is validated

|{cgu-operator-first}
|version 4.13

|Server hardware
|3 x Dell PowerEdge R650 rack servers

|NVMe hard disks
a|* 50 GB disk for `/var/lib/etcd`
* 2.9 TB disk for `/var/lib/containers`

|SSD hard disks
a|* 1 SSD split into 15 200GB thin-provisioned logical volumes provisioned as `PV` CRs
* 1 SSD serving as an extra large `PV` resource

> |====

Reviewer 3: The above considerations talk about number of policies. I think we should include here the number of policies used to define the DU configuration being applied/managed. It should be 5 policies (6 if you include a validation policy). @akrzos can confirm.

Reviewer: 5 policies total.  In the scale lab we test with the typical 3 policies plus two policies are renamed because there would be a site policy and there would be a du validation policy.

[IMPORTANT]
====

> The following network specifications are based on a simulated lab environment and represent best guess data points for a real-world RAN network.

Reviewer 3: ```suggestion
The following network specifications were selected as representative of a typical real-world RAN network and were applied to the scale lab environment during testing.
```

====

.Simulated managed cluster network specifications
[cols=2*, width="90%", options="header"]
|====
|Specification
|Description

|Round-trip time (RTT) latency

> |50ms RTT measured in the lab environment, representative of real-world network edge environment

Reviewer 3: I think the "important" covers the context and the 50ms was what was explicitly implemented in the lab network (rather than just being measured).
```suggestion
|50ms
```

|Packet loss

> |0.02% packet loss in the lab environment

Reviewer 3: ```suggestion
|0.02% packet loss
```

|Lab bandwidth tests
a|* 20 Mbps
* 100 Mbps

> * Unlimited

Reviewer 2: I don't understand this.
What is unlimited bandwidth?
And at which bandwidth was managing 3500 SNOs achieved?

// Module included in the following assemblies:
//
// * scalability_and_performance/ztp_far_edge/ztp-preparing-the-hub-cluster.adoc]

:_content-type: REFERENCE
[id="ztp-gitops-ztp-max-spoke-clusters_{context}"]
= Recommended hub cluster specifications and managed cluster limits for {ztp}

With {ztp-first}, you can manage thousands of clusters in geographically dispersed regions and networks.

> The Red Hat Performance and Scale lab successfully created and managed 3500 virtual {sno} clusters with a reduced DU profile from a single {rh-rhacm-first} hub cluster in a lab environment.footnote:[link:https://access.redhat.com/documentation/en-us/red_hat_advanced_cluster_management_for_kubernetes/2.7/html/install/installing#single-node[Creating and managing {sno} clusters with {rh-rhacm}]]

Reviewer 4: The repo guidelines say to avoid footnotes: https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#footnotes

I think you could add this to the additional resources section

> In real world situations, the scaling limits for the number of clusters that you can manage will vary depending on various factors affecting the hub cluster.

Reviewer 4: You have real world  and real-world listed in this doc and to be honest I am not sure which is the correct one.

Author: I think we should hyphenate "real world" when it is used as an adjective. So "real-world" here.

For example:

> Hub cluster resources::

Reviewer 4: Based on the mod doc guide, I wonder if these should be discrete headings https://redhat-documentation.github.io/modular-docs/#reference-module-guidelines

I have never seen a list like this before and didn't see it called out in the style guide. I could have missed it if it is in there.

Author: It's a description list. See here for more details: https://docs.asciidoctor.org/asciidoc/latest/lists/description/
